alg/viewshed.cpp

```diff
@@ -426,499 +480,389 @@ bool Viewshed::run(GDALRasterBandH hBand, GDALProgressFunc pfnProgress,
 
-            setVisibility(iPixel, dfZ, padfFirstLineVal, vResult);
-        }
-        else
+    std::thread t1(
+        [&, left = iLeft]()
         {
-            for (; iPixel >= 0; iPixel--)
-            {
-                pabyResult[iPixel] = oOpts.outOfRangeVal;
-                if (oOpts.outputMode != OutputMode::Normal)
-                    dfHeightResult[iPixel] = oOpts.outOfRangeVal;
-            }
-        }
-    }
-    /* process right direction */
-    for (int iPixel = nX + 2; iPixel < nXSize; iPixel++)
-    {
-        dfGroundLevel = 0;
-        if (oOpts.outputMode == OutputMode::DEM)
-            dfGroundLevel = padfFirstLineVal[iPixel];
-        if (AdjustHeightInRange(adfGeoTransform.data(), iPixel - nX, 0,
-                                padfFirstLineVal[iPixel], dfDistance2,
-                                oOpts.curveCoeff, dfSphereDiameter))
+            processLineLeft(nX, nYOffset, nX - 2, left - 1, vResult,
+                            vThisLineVal, vLastLineVal);
+        });
+
+    std::thread t2(
+        [&, right = iRight]()
         {
-            double dfZ = CalcHeightLine(
-                iPixel - nX, padfFirstLineVal[iPixel - 1], dfZObserver);
+            processLineRight(nX, nYOffset, nX + 2, right, vResult, vThisLineVal,
+                             vLastLineVal);
+        });
+    t1.join();
+    t2.join();
 
-            if (oOpts.outputMode != OutputMode::Normal)
-                dfHeightResult[iPixel] = std::max(
-                    0.0, (dfZ - padfFirstLineVal[iPixel] + dfGroundLevel));
+    // Make the current line the last line.
+    vLastLineVal = std::move(vThisLineVal);
 
-            setVisibility(iPixel, dfZ, padfFirstLineVal, vResult);
-        }
-        else
-        {
-            for (; iPixel < nXSize; iPixel++)
-            {
-                pabyResult[iPixel] = oOpts.outOfRangeVal;
-                if (oOpts.outputMode != OutputMode::Normal)
-                    dfHeightResult[iPixel] = oOpts.outOfRangeVal;
-            }
-        }
-    }
-    /* write result line */
+    // Create the output writer.
+    if (!writeLine(nY, vResult))
+        return false;
 
-    void *data;
-    GDALDataType dataType;
-    if (oOpts.outputMode == OutputMode::Normal)
+    if (!lineProgress())
+        return false;
+    return true;
+}
+
+/// Process a line above or below the observer.
+///
+/// @param nX  X location of the observer
+/// @param nY  Y location of the observer
+/// @param nLine  Line number being processed.
+/// @param vLastLineVal  Vector in which to store the read line. Becomes the last line
+///    in further processing.
+/// @return True on success, false otherwise.
+bool Viewshed::processLine(int nX, int nY, int nLine,
+                           std::vector<double> &vLastLineVal)
+{
+    int nYOffset = nLine - nY;
+    std::vector<double> vResult(oOutExtent.xSize());
+    std::vector<double> vThisLineVal(oOutExtent.xSize());
```

Comment:
isn't there a runtime cost in allocating those temporary buffers every time this method is invoked ?

Reply:
I'm sure there is. But it's small. Probably very small. malloc is very good and you're re-allocating vectors of the same size that you just recently destroyed. Also, from a code standpoint, it means that you don't have to worry about making sure that threads maintain local copies. I ran things for time after each change, and this made no noticeable difference in my test, but there are lots of cases and I won't swear that it won't make a difference for someone else's test case.